File: 0022_Generate_Parentheses/Generate_Parentheses.rb
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  result = []
  generate(result, '', 0, 0, n)
  result
end

private

# @param {String[]} result
# @param {String} str
# @param {Integer} left
# @param {Integer} right
# @param {Integer} n
def generate(result, str, left, right, n)
  if str.length == 2 * n
    result.push(str)
    return
  end
  generate(result, str + '(', left + 1, right, n) if left < n
  generate(result, str + ')', left, right + 1, n) if right < left
end

File: 0022_Generate_Parentheses/Generate_Parentheses.py
# -*- coding: utf-8 -*-

"""
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
"""


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self._generate(result, '', 0, 0, n)
        return result

    def _generate(self, result: List[str], s: str, left: int, right: int, n: int) -> None:
        if len(s) == 2 * n:
            result.append(s)
            return
        if left < n:
            self._generate(result, s + '(', left + 1, right, n)
        if right < left:
            self._generate(result, s + ')', left, right + 1, n)
